### PR TITLE
Shortcut to inject the editor

### PR DESCRIPTION
**Pull request by dmarcos**

### CODE REVIEW

File: src/components/scene/editor.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    if (!this.data || this.editorInjected || !shortcutPressed) { return; }
+    this.injectEditor();
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
+    script.type = 'text/javascript';
+    script.async = true;
+    script.onload = function () { console.log('A-Frame Editor Loaded'); };
+    script.src = 'https://fernandojsg.github.io/aframe-editor/build/aframe-editor.js';
```
Comment (cvan) [inline src/components/scene/editor.js]: can we have a const defined at the top of this file?

File: src/components/scene/editor.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    window.addEventListener('keydown', this.onKeydown, false);
+  },
+
+  pause: function () {
+    window.removeEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
```
Comment (cvan) [inline src/components/scene/editor.js]: we have [this helper function](https://github.com/aframevr/aframe/blob/f0c2b9f/src/utils/index.js#L126-L136). either use that, or check `document.activeElement` or make sure that `evt.target` is not a form field, `[contexteditable]` element, etc.

File: src/components/scene/editor.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    this.injectEditor();
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
+    script.type = 'text/javascript';
+    script.async = true;
+    script.onload = function () { console.log('A-Frame Editor Loaded'); };
+    script.src = 'https://fernandojsg.github.io/aframe-editor/build/aframe-editor.js';
+    document.getElementsByTagName('head')[0].appendChild(script);
```
Comment (cvan) [inline src/components/scene/editor.js]: you can just use `document.head`

File: src/components/scene/editor.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    this.editorInjected = true;
```
Comment (cvan) [inline src/components/scene/editor.js]: technically not true [yet](https://github.com/aframevr/aframe/pull/1599/files#diff-565a020f4c51daed65e6dca6c67d7e8eR29), but I [understand](https://github.com/aframevr/aframe/pull/1599/files#diff-565a020f4c51daed65e6dca6c67d7e8eR21) why you've done this.

File: src/components/scene/editor.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  },
+
+  pause: function () {
+    window.removeEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
+    if (!this.data || this.editorInjected || !shortcutPressed) { return; }
```
Comment (cvan) [inline src/components/scene/editor.js]: what if `this.data.enabled === false`?

File: src/components/scene/editor.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    this.injectEditor();
```
Comment (cvan) [inline src/components/scene/editor.js]: if I have multiple scenes on the page, the Editor script will still get injected again

File: src/components/scene/editor.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
+    script.type = 'text/javascript';
+    script.async = true;
```
Comment (cvan) [inline src/components/scene/editor.js]: [> Scripts that are dynamically created and added to the document are async by default](http://www.html5rocks.com/en/tutorials/speed/script-loading/)

File: src/components/scene/editor.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
+    if (!this.data || this.editorInjected || !shortcutPressed) { return; }
+    this.injectEditor();
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
+    script.type = 'text/javascript';
```
Comment (cvan) [inline src/components/scene/editor.js]: can remove

File: src/components/scene/editor.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    script.async = true;
+    script.onload = function () { console.log('A-Frame Editor Loaded'); };
```
Comment (cvan) [inline src/components/scene/editor.js]: shouldn't the Editor do its logging? (and perhaps spit out `A-Frame Editor Version: 0.1.0`?)

File: src/components/scene/editor.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+module.exports.Component = registerComponent('editor', {
+  schema: { default: true },
+
+  init: function () {
+    this.onKeydown = this.onKeydown.bind(this);
+  },
+
+  play: function () {
+    window.addEventListener('keydown', this.onKeydown, false);
```
Comment (cvan) [inline src/components/scene/editor.js]: `false` is the default

File: src/components/scene/editor.js (hunk 1 above)
Comment (cvan) [inline src/components/scene/editor.js]: or in `src/index.js`?

File: src/components/scene/editor.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        AFRAME.editorInjected || !shortcutPressed) { return; }
+    this.injectEditor();
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
+    script.type = 'text/javascript';
+    script.src = EDITOR_URL;
+    document.getElementsByTagName('head')[0].appendChild(script);
+    AFRAME.editorInjected = true;
```
Comment (ngokevin) [inline src/components/scene/editor.js]: Does this need to be global rather than `this.injected`?

File: src/components/scene/editor.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
+    if (document.activeElement !== document.body || !this.data ||
+        AFRAME.editorInjected || !shortcutPressed) { return; }
+    this.injectEditor();
+  },
+
+  injectEditor: function () {
+    var script = this.script = document.createElement('script');
```
Comment (ngokevin) [inline src/components/scene/editor.js]: Doesn't look like `this.script` is used.

File: src/components/scene/editor.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
+/* global AFRAME */
+var registerComponent = require('../../core/component').registerComponent;
+
+var EDITOR_URL = 'https://aframe.io/aframe-editor/build/aframe-editor.js';
```
Comment (ngokevin) [inline src/components/scene/editor.js]: We should version this later `https://aframe.io/aframe-editor/builds/0.3.0/aframe-editor.js`

**Comment by ngokevin** [comment]
r+wc

**Comment by dmarcos** [comment]
This is ready to go but I first need @fernandojsg input first. In order to lock the version of the editor to a version of A-Frame we have to agree on the url schema. @ngokevin suggested `https://aframe.io/aframe-editor/builds/0.3.0/aframe-editor.js` where the version refers to the A-Frame version. I made the editor url a property of the component so people filling adventurous can try experimental versions

**Comment by fernandojsg** [comment]
Yep, I agree that probably will be more useful to follow the same version number for the builds to match aframe. We could use `https://aframe.io/aframe-editor/builds/0.3.0/aframe-editor.js` for the first release

**Comment by ngokevin** [comment]
Alternatively `https://aframe.io/releases/0.3.0/aframe-editor.min.js` to match A-Frame.

**Comment by fernandojsg** [comment]
@ngokevin :+1:

**Comment by dmarcos** [comment]
@ngokevin @fernandojsg How are we going to deal with master? If you use a-frame master it should point to the master version of the editor as well.

**Comment by ngokevin** [comment]
Maybe we can immediately point A-Frame Master to Editor Master after we release. But A-Frame 0.3.0 should be locked to Editor 0.3.0.

Or we always point to the last release. If developers are using Master, we prescribe to configure their Editor URL. `<a-scene editor="url: mastereeditorurl">`

**Comment by dmarcos** [comment]
@ngokevin are you want to add the release command for the editor? As you mentioned master should point to the latest release of the editor and devs can just change the url if they want bleeding edge. Before merging this It would great to have a functional URL `ttps://aframe.io/aframe-editor/builds/0.3.0/aframe-editor.js`

**Comment by ngokevin** [comment]
How about I copy one to `https://aframe.io/releases/0.3.0/aframe-editor.min.js`?

**Comment by dmarcos** [comment]
Yes we can copy it manually for now.

**Comment by fernandojsg** [comment]
Remember to `s/editor/inspector`

**Comment by dmarcos** [comment]
Let's push a build to `https://aframe.io/releases/0.3.0/aframe-inspector.min.js` so I can point this PR to it.

File: src/components/scene/editor.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  play: function () {
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  pause: function () {
+    window.removeEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
```
Comment (ngokevin) [inline src/components/scene/editor.js]: We should change this to `i` for Inspector.

**Comment by ngokevin** [comment]
`ctrl + alt + i/e` loads the editor and immediately jumps into it the first time.

When I exit, and press `ctrl + alt + i/e` again, I expected it to open the Editor again, just like the Browser DOM Inspector you can toggle on/off with the same shortcut.

File: src/components/scene/editor.js (hunk 14 above)
Comment (ngokevin) [inline src/components/scene/editor.js]: Also nice because it matches Chrome's `cmd + alt + i` shortcut.

**Comment by dmarcos** [comment]
In Windows the shortcut is `ctrl+shift+i` . `windows key +alt+e` opens the browser menu in chrome.

**Comment by dmarcos** [comment]
@fernandojsg how can I programmatically exit the editor so I can close/open it with the shortcut? Maybe the shortcut should be handled by the editor instead of aframe. What do you think?

**Comment by fernandojsg** [comment]
@dmarcos once the editor is injected you can use `editor.enable()` and `editor.disable()` (I just realised that I should change also s/editor/inspector there :)
I could include the shortcut in the inspector to toggle it, but to inject the first time should be done outside the editor, how to handle this two entry points?

**Comment by ngokevin** [comment]
Just a nitpick, could be `editor.open()/editor.close()` or `editor.enter()/enter.exit()`.

**Comment by fernandojsg** [comment]
@ngokevin yep, just after writing that, I was thinking how bad naming these functions have. I'll change them for `open/close`

File: src/components/scene/inspector.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  play: function () {
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  pause: function () {
+    window.removeEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
```
Comment (ngokevin) [inline src/components/scene/inspector.js]: `alt + ctrl + i` ?

File: src/components/scene/inspector.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  toggleInspector: function () {
+    if (AFRAME.inspector.opened) {
+      AFRAME.inspector.close();
+    } else {
+      AFRAME.inspector.open();
+    }
+  },
+
+  injectInspector: function () {
+    var script = document.createElement('script');
```
Comment (ngokevin) [inline src/components/scene/inspector.js]: Declare before logic. So we don't unnecessarily create a script element if the inspector isn't injected.

**Comment by ngokevin** [comment]
r+wc

File: src/components/scene/inspector.js (hunk 15 above)
Comment (fernandojsg) [inline src/components/scene/inspector.js]: Yep probably that will match better for `inspector` as similar as devtools, and leave `e` for the incoming VR-editor

**Comment by fernandojsg** [comment]
One question regarding this shortcut. If you press `ctrl+alt+i` to inject the `inspector` and open it, should I add also the same shortcut once injected to toggle the inspector right? (Even if after that we could create a easier shortcut like `Esc` or whatever for doing the same)

**Comment by ngokevin** [comment]
Yeah, would be nice to toggle like the Chrome Devtools.

**Comment by FrederickDesimpel** [comment]
just wondering... since these are the same shortcuts... will this replace the browsers inspector? can' t we have both?

**Comment by ngokevin** [comment]
I think we're using alt instead of shift.

**Comment by cvan** [comment]
> I think we're using alt instead of shift.

nice, ok - can we update the docs?

File: src/components/scene/inspector.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+  init: function () {
+    this.onKeydown = this.onKeydown.bind(this);
+  },
+
+  play: function () {
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  pause: function () {
+    window.removeEventListener('keydown', this.onKeydown);
```
Comment (fernandojsg) [inline src/components/scene/inspector.js]: When the editor starts, it pause the scene, causing the shortcut to be removed (https://github.com/aframevr/aframe-inspector/issues/257).

File: package.json [hunk 18]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "aframe",
-  "version": "0.2.0",
+  "version": "0.3.0",
```
Comment (ngokevin) [inline package.json]: I'd probably move this out to its own commit, but same PR

File: src/components/scene/inspector.js [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
```
Comment (cvan) [inline src/components/scene/inspector.js]: there might be some form field on the page. so this should be checking if we should be listening for the keypress. you can either use [our `utils.shouldCaptureKeyEvent` method](https://github.com/aframevr/aframe/blob/master/src/utils/index.js#L127-L137).

what was that case when you said `document.activeElement` was not `<body>` in some browser in some case? I can't reproduce that issue you were talking about.

File: src/components/scene/inspector.js [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+    }
+  },
+
+  injectInspector: function () {
+    var script = document.createElement('script');
+    if (AFRAME.inspectorInjected) { return; }
+    script.type = 'text/javascript';
+    script.src = this.data.url;
+    document.getElementsByTagName('head')[0].appendChild(script);
+    AFRAME.inspectorInjected = true;
```
Comment (cvan) [inline src/components/scene/inspector.js]: why is this not `AFRAME.inspector.injected`?

File: src/components/scene/inspector.js [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+      AFRAME.inspector.open();
+    }
+  },
+
+  injectInspector: function () {
+    var script = document.createElement('script');
+    if (AFRAME.inspectorInjected) { return; }
+    script.type = 'text/javascript';
+    script.src = this.data.url;
+    document.getElementsByTagName('head')[0].appendChild(script);
```
Comment (cvan) [inline src/components/scene/inspector.js]: `document.head` works fine

File: src/components/scene/inspector.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+      AFRAME.inspector.close();
+    } else {
+      AFRAME.inspector.open();
+    }
+  },
+
+  injectInspector: function () {
+    var script = document.createElement('script');
+    if (AFRAME.inspectorInjected) { return; }
+    script.type = 'text/javascript';
```
Comment (cvan) [inline src/components/scene/inspector.js]: don't need

File: src/components/scene/inspector.js [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
+    var shortcutPressed = evt.keyCode === 69 && evt.ctrlKey && evt.altKey;
+    if (!this.data || !shortcutPressed) { return; }
+    this.injectInspector();
+    this.toggleInspector();
+  },
+
+  toggleInspector: function () {
+    if (AFRAME.inspector.opened) {
```
Comment (cvan) [inline src/components/scene/inspector.js]: should a method perhaps live in the Inspector repo called `toggle`?

File: src/components/scene/inspector.js (hunk 19 above)
Comment (dmarcos) [inline src/components/scene/inspector.js]: In the case of embedded scenes the activeElement might not be the body. The editor widgets can capture the focus and you still want the shortcut to work.

File: src/components/scene/inspector.js (hunk 23 above)
Comment (cvan) [inline src/components/scene/inspector.js]: it's possible that this code gets run before the `<script>` is actually injected by `injectInspector`.

even checking for `AFRAME.inspectorInjected` isn't enough. you'll want to add a callback or Promise for the `script.onload` for `injectInspector`.

File: src/components/scene/inspector.js (hunk 20 above)
Comment (dmarcos) [inline src/components/scene/inspector.js]: inspector is the scope where the inspector inject itself. inspectorInjected is the flag that aframe uses that the inspector has been already injected and we don't load it multiple times.

File: src/components/scene/inspector.js (hunk 16 above)
Comment (cvan) [inline src/components/scene/inspector.js]: good call

File: src/components/scene/inspector.js [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+    var script = document.createElement('script');
+    if (AFRAME.inspectorInjected) { return; }
+    script.type = 'text/javascript';
+    script.src = this.data.url;
+    document.getElementsByTagName('head')[0].appendChild(script);
+    AFRAME.inspectorInjected = true;
+  },
+
+  remove: function () {
+    window.removeEventListener('keyup', this.listener);
```
Comment (cvan) [inline src/components/scene/inspector.js]: `this.listener` isn't actually defined above

**Comment by dmarcos** [comment]
I suggest to not do reviews if the label is not set to `review` this PR only went through a rebase cycle

File: src/components/scene/inspector.js (hunk 19 above)
Comment (dmarcos) [inline src/components/scene/inspector.js]: The shouldCaptureKeyEvent also returns if the alt or ctrl keys are pressed. We are capturing `ctrl + alt + e`

File: src/components/scene/inspector.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    url: {default: INSPECTOR_URL}
+  },
+
+  init: function () {
+    this.onKeydown = this.onKeydown.bind(this);
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + e
```
Comment (ngokevin) [inline src/components/scene/inspector.js]: Should be `ctrl + alt + i`

**Comment by ngokevin** [comment]
r+

can do later:

- unit tests
- bump package.json
- test with most recent inspector build that ctrl/alt/i can toggle inspector open and closed

**Comment by mkungla** [comment]
restarted the build

File: src/components/scene/inspector.js (hunk 19 above)
Comment (cvan) [inline src/components/scene/inspector.js]: in the case of embedded scenes, the `<iframe>` can't listen to any keypresses until you first click it. on page load, the `document.activeElement` will be `<body>` of the parent window.

and, it's probably rare, but the keyboard shortcut might get accidentally triggered if a user typed it while typing in a form field (e.g., a `<textarea>` on a Wikipedia page or a comment field of a page with embedded scenes made with A-Frame).

also, what happens on the page when I have two embedded scenes? we probably don't want both listeners to fire?

File: src/components/scene/inspector.js [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+    var script;
+    if (AFRAME.inspectorInjected) { return; }
+    script = document.createElement('script');
+    script.src = this.data.url;
+    document.head.appendChild(script);
+    AFRAME.inspectorInjected = true;
+  },
+
+  remove: function () {
+    window.removeEventListener('keydown', this.keydown);
```
Comment (cvan) [inline src/components/scene/inspector.js]: typo: should be `this.onKeydown`

File: src/components/scene/inspector.js [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+
+  init: function () {
+    this.onKeydown = this.onKeydown.bind(this);
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + i
+    var shortcutPressed = evt.keyCode === 73 && evt.ctrlKey && evt.altKey;
+    var escape = evt.keyCode === 27;
```
Comment (cvan) [inline src/components/scene/inspector.js]: IMO, this should also check `evt.target`. we're assuming that the scene won't be listening for Escape at all? that's a pretty common keypress to listen for.

**Comment by cvan** [comment]
just marked `needs changes` for the typo in the `remove` method

File: src/components/scene/inspector.js (hunk 27 above)
Comment (ngokevin) [inline src/components/scene/inspector.js]: Code is a little misleading because it only applies when the Inspector is open, which the scene shouldn't be listening to anything at that point.

Comment (dmarcos) [inline src/components/scene/inspector.js]: Yes. The escape shortcut is to close the editor that will be occupying the whole screen with the scene paused. The editor should be the only thing in the page listening for key presses.

File: src/components/scene/inspector.js [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+  init: function () {
+    this.onKeydown = this.onKeydown.bind(this);
+    window.addEventListener('keydown', this.onKeydown);
+  },
+
+  onKeydown: function (evt) {
+    // Alt + Ctrl + i
+    var shortcutPressed = evt.keyCode === 73 && evt.ctrlKey && evt.altKey;
+    if (!this.data || !shortcutPressed) { return; }
+    this.injectInspector();
```
Comment (dmarcos) [inline src/components/scene/inspector.js]: I simplified this. The component just injects the editor the first time `ctrl + alt + i` and then the editor takes over listening to keyboard events.
